runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -294,218 +269,258 @@ pub mod pallet {
 
-			let entropy = compute_entropy::<T>(parent_hash);
-			let mut rng = rand_chacha::ChaChaRng::from_seed(entropy.into());
+		let dispute_statement_set_valid = move |set: DisputeStatementSet| {
+			T::DisputesHandler::filter_dispute_data(set, post_conclusion_acceptance_period)
+		};
 
-			let (checked_disputes, checked_disputes_weight) = limit_and_sanitize_disputes::<T, _>(
+		// Limit the disputes first, since the following statements depend on the votes include here.
+		let (checked_disputes_sets, checked_disputes_sets_consumed_weight) =
+			limit_and_sanitize_disputes::<T, _>(
 				disputes,
-				&dispute_set_validity_check,
+				dispute_statement_set_valid,
 				max_block_weight,
 				&mut rng,
 			);
-			(
-				checked_disputes,
-				checked_disputes_weight
-					.saturating_add(candidates_weight)
-					.saturating_add(bitfields_weight),
-			)
-		};
 
-		let expected_bits = <scheduler::Pallet<T>>::availability_cores().len();
+		// Assure the maximum block weight is adhered, by limiting bitfields and backed
+		// candidates. Dispute statement sets were already limited before.
+		let actual_weight = apply_weight_limit::<T>(
+			&mut backed_candidates,
+			&mut bitfields,
+			max_block_weight.saturating_sub(checked_disputes_sets_consumed_weight),
+			&mut rng,
+		);
 
-		// Handle disputes logic.
-		let disputed_bitfield = {
-			let new_current_dispute_sets: Vec<_> = checked_disputes
-				.iter()
-				.map(AsRef::as_ref)
-				.filter(|s| s.session == current_session)
-				.map(|s| (s.session, s.candidate_hash))
-				.collect();
+		if actual_weight.any_gt(max_block_weight) {
+			log::warn!(target: LOG_TARGET, "Post weight limiting weight is still too large.");
+		}
 
-			// Note that `process_checked_multi_dispute_data` will iterate and import each
-			// dispute; so the input here must be reasonably bounded,
-			// which is guaranteed by the checks and weight limitation above.
-			let _ = T::DisputesHandler::process_checked_multi_dispute_data(&checked_disputes)?;
-			METRICS.on_disputes_imported(checked_disputes.len() as u64);
+		// Note that `process_checked_multi_dispute_data` will iterate and import each
+		// dispute; so the input here must be reasonably bounded,
+		// which is guaranteed by the checks and weight limitation above.
+		// We don't care about fresh or not disputes
+		// this writes them to storage, so let's query it via those means
+		// if this fails for whatever reason, that's ok.
+		if let Err(e) =
+			T::DisputesHandler::process_checked_multi_dispute_data(&checked_disputes_sets)
+		{
+			log::warn!(target: LOG_TARGET, "MultiDisputesData failed to update: {:?}", e);
+		};
+		METRICS.on_disputes_imported(checked_disputes_sets.len() as u64);
 
-			if T::DisputesHandler::is_frozen() {
-				// Relay chain freeze, at this point we will not include any parachain blocks.
-				METRICS.on_relay_chain_freeze();
+		set_scrapable_on_chain_disputes::<T>(current_session, checked_disputes_sets.clone());
```

Comment:
We also don't seem to care about this weight .

Reply:
Same reason.